materialize-boilerplate/info_schema.go

```diff
@@ -0,0 +1,211 @@
+	_, ok := i.resources[rk]
+	return ok
+}
+
+// ExtractProjection finds the projection in a Flow collection spec that must correspond to a field
+// reported by the endpoint. Fields in the collection spec must have the same one-way translation
+// applied to them that a field does when materializing it in order to do this lookup.
+func (i *InfoSchema) ExtractProjection(efn string, collection pf.CollectionSpec) (pf.Projection, bool, error) {
+	var found bool
+	var out pf.Projection
+
+	for idx := range collection.Projections {
+		matches := i.translateField(collection.Projections[idx].Field) == efn
+		if found && matches {
+			// This should never happen since the standard constraints from `Validator` forbid it,
+			// but I'm leaving it here as a sanity check just in case.
+			return pf.Projection{}, false, fmt.Errorf("ambiguous endpoint field name when looking for projection %q", efn)
+		} else if matches {
+			found = true
+			out = collection.Projections[idx]
+		}
+	}
+
+	return out, found, nil
+}
+
+// InSelectedFields returns true if the provided endpoint field name (as reported by the endpoint)
+// exists in the field selection for a materialization. It does a similar translation of fields from
+// the field selection as ExtractProjection in order to do this matching.
+func (i *InfoSchema) InSelectedFields(efn string, fs pf.FieldSelection) (bool, error) {
+	var found bool
+
+	for _, f := range fs.AllFields() {
+		matches := i.translateField(f) == efn
+		if found && matches {
+			// This should never happen since the standard constraints from `Validator` forbid it,
+			// but I'm leaving it here as a sanity check just in case.
+			return false, fmt.Errorf("ambiguous endpoint field name when looking for selected field %q", efn)
+		} else if matches {
+			found = true
+		}
+	}
+
+	return found, nil
+}
+
+// AmbiguousResourcePaths is a utility for determining if a resource path could be ambiguous when
+// materialized to the destination system. This should only be possible if the destination system
+// (or the connector) transforms resource paths in some way that could result in conflicts. Usually
+// the connector should take this into consideration when producing the resource path with its
+// `Validate` response, but we haven't always done a good job with that.
+func (i *InfoSchema) AmbiguousResourcePaths(resourcePaths [][]string) [][]string {
+	seenPaths := make(map[string][][]string)
+
+	for _, rp := range resourcePaths {
+		rk := joinPath(i.locatePath(rp))
+		seenPaths[rk] = append(seenPaths[rk], rp)
+	}
+
+	var out [][]string
+	for _, p := range seenPaths {
+		if len(p) > 1 {
+			out = append(out, p...)
+		}
+	}
+
+	// Sort the output to make it a little more clear which paths are ambiguous if there are a lot
+	// of them.
+	slices.SortFunc(out, func(a, b []string) int {
+		return strings.Compare(joinPath(i.locatePath(a)), joinPath(i.locatePath(b)))
+	})
+
+	return out
+}
+
+func joinPath(in []string) string {
```

Comment:
Why not just use https://pkg.go.dev/strings#Join?

Reply:
I may have had some grander plan for this at some earlier iteration, but yeah that's all we'd need now, and I'll swap that in!